DietPi-Software | AdGuard Home

Joulinar opened the pull request:
AdGuard Home - add new software title. Software **ID 126**


**Status**: testing
- [x] AdGuard Home install
- [x] AdGuard Home uninstall
- [x] AdGuard Home reinstall
- [x] install/uninstall/reinstall together with Unbound
- [x] prevent install/uninstall/reinstall if Pi-hole is present

**test device**
- [x] RPi1
- [x] RPi3B+ 32bit
- [x] RPi4B 64bit
- [x] VM amd64 Buster
- [x] VM amd64 Stretch

**Reference**: https://github.com/MichaIng/DietPi/issues/3778

**Commit list/description**:
- DietPi-Software |  AdGuard Home - add new software title. create initial configuration
- DietPi-Services | AdGuard Home - add new service 
- DietPi-Software | AdGuard Home - stop service on reinstall before copying files
- DietPi-Software | AdGuard Home - adjust product name
- DietPi-Software | AdGuard Home - change software ID to 126
- DietPi-Services | AdGuard Home - adjust product name
- README.md | AdGuard Home - add new software title
- DietPi-Software | AdGuard Home - change configuration file ID
- DietPi-Software | AdGuard Home - adjust service configuration file
- AdGuardHome.yaml | remove non-default DNS filter lists
- DietPi-Software | AdGuard Home - remove logfiles on uninstall from /var/log/
- DietPi-Software | AdGuard Home - rename service and keep output on STDOUT (default)
- DietPi-Services | AdGuard Home - rename service
- AdGuardHome.yaml | change default port to 8083
- DietPi-Software | AdGuard Home - remove environment configuration from service description
- DietPi-Software | AdGuard Home - Avoid having AdGuard Home installed in parallel to Pi-hole
- DietPi-Software | AdGuard Home - Offer to install Unbound along with AdGuard Home
- DietPi-Software | Unbound - configure Unbound to work together with AdGuard Home
- DietPi-Software | Unbound - Configuring AdGuard Home to use Unbound
- DietPi-Software | AdGuard Home - Unbound: Switch port to 5335 if it was installed before
- DietPi-Software | AdGuard Home - change configuration to use Unbound if available
- DietPi-Software | Unbound - Assure AdGuard Home does not resolve via Unbound anymore if Unbound will be uninstalled
- DietPi-Software | Pi-hole – on uninstallation, switch Unbound port to 53 if Unbound is still installed
- DietPi-Software | AdGuard Home – on uninstallation, switch Unbound port to 53 if Unbound is still installed
- DietPi-Software | AdGuard Home - ensure no concurrent installation of AdGuard Home and Pi-hole, even if installation is started from CLI
- DietPi-Software | Unbound - Show a notification only, instead of a message pop-up if Unbound is being uninstalled
- DietPi-Software | AdGuard Home: Do Pi-hole concurrency check as part of the dependency function, like we do with Tor Relay and Tor Hotspot. This is called on CLI installs as well while on menu-based installs the additional warnings are great so that users have chance to adjust their selections before starting the installs.
- DietPi-Software | AdGuard Home: Assure that file modes are kept when copying files to disk and remove files from tmpfs afterwards
- DietPi-Software | AdGuard Home: Only enable service during install and start it after all installs have been finished, to keep RAM and CPU usage at a minimum. If it was installed already, the service is still running, which is then required to avoid stopping the systems own DNS server.
- DietPi-Software | Assure that uninstalls do not fail uncleanly when single steps fail, but keep the interactive error prompt. This way it is made clear that something failed, but instead of "Exit", "Ignore" can be selected to proceed with the uninstall.
- DietPi-Software | Lighttpd - stop service before uninstalling, otherwise an active process will be left
- DietPi-Software | Lighttpd – remove enabled configuration directory, to avoid other software title trying to disable Lighttpd mods
- DietPi-Software | Pi-hole – disable Lighttpd mods one by one as using wildcards does not seems to be working
- DietPi-Software | Pi-hole – correct folder name for available Lighttpd mods to allow correct removal during uninstallation
- DietPi-Software | Lighttpd - enhanced Lighttpd uninstall process
- DietPi-Software | AdGuard Home - Allow AdGuardHome to run without superuser privileges
- DietPi-Software | AdGuard Home - Allow AdGuardHome to run without superuser privileges via service file value `AmbientCapabilities`
- AdGuardHome.yaml | reduce query log interval down to 7 days


---

ok first shot. Quite some thinks to be clarified

**done**
- [x] which directory to run on? 
AdGuard Home will be located in `/opt`
- [x] default web server port?
AdGuard Home default web server port will be `8083`
- [x] should we pre-check for already installed PiHole/Unbound?
pre-check are done during software selection
if available, Unbound/AdGuard Home will be reconfigured to be able to work together
- [x] which user to be used on Web UI? 
user `admin` will be used to login to Web UI
- [x] configuration
logs are available via `journalctl`
- [x] running AGH without superuser privileges
done via service file value `AmbientCapabilities`
- [x] service file configuration
`restart` set to `on-failure` and set restart attempt to 5 times only
- [x] tune default config parameter?
done

**Info**
Web UI pw = global pw is used (requires `apache2-utils`)

Joulinar committed:
v7.3

DietPi-Software |  AdGuardHome - add new software title. create initial configuration

Joulinar committed:
v7.3

DietPi-Software | AdGuardHome - stop service on reinstall before copying files

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1421,6 +1421,13 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
+		
+		software_id=171
+
+		aSOFTWARE_NAME[$software_id]='AdGuardHome'
```
Let's spell it "AdGuard Home" with the space, matching official spelling: https://adguard.com/de/adguard-home/overview.html

Joulinar replied:
done

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8966,6 +9010,43 @@ _EOF_
+
+[Service]
+StartLimitInterval=5
+StartLimitBurst=10
```
These two need to go to the `[Unit]` section and `StartLimitBurst` is now called `StartLimitIntervalSec`: https://www.freedesktop.org/software/systemd/man/systemd.unit.html#StartLimitIntervalSec=interval

The question is whether we want this at all. Is there some reason why a start can fail while an immediate second attempt will succeed? Or for `Restart=always`, is there an internal updater or such that triggers a restart?

Joulinar replied:
for the initial setup, I used the original AGH service definition. Seems they are not up-to-date

Joulinar replied:
done

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8966,6 +9010,43 @@ _EOF_
+StandardError=file:/var/log/AdGuardHome.err
+Restart=always
+RestartSec=10
+EnvironmentFile=-/etc/sysconfig/AdGuardHome
```
`/etc/sysconfig` is more a CentOS-ish standard, while on Debian-based distros `/etc/default` is used instead. But of course this only makes sense if AdGuard Home internally uses environment variables or when we pass them as command line options (which is currently not the case).

Joulinar replied:
ok I changed it but it doesn't seems to be used

MichaIng replied:
I also looked into the [docs](https://github.com/AdguardTeam/AdGuardHome/wiki/Configuration) but it doesn't seem to take care of environment variables. In this case, I'd just remove that part. Theoretically we could do something like:
```
EnvironmentFile=-/etc/default/adguardhome
ExecStart=/opt/AdGuardHome/AdGuardHome -s start $CMD_ARGS
```
and then allow users to set `CMD_ARGS="something"` in `/etc/default/adguardhome`. But from what I see, all important settings can be set via the YAML configuration file already. If someone really wants to e.g. disable update checks, `dietpi-services` can be used to create a service override config.

Joulinar replied:
ok I removed the `EnvironmentFile` for now

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -1421,6 +1421,13 @@ INDEX_BROWSER_TARGET=$INDEX_BROWSER_TARGET"
 		aSOFTWARE_CATX[$software_id]=13
 		aSOFTWARE_DOCS[$software_id]='https://dietpi.com/docs/software/dns_servers/#unbound'
+		
+		software_id=171
```
ID conflict with [frp](https://github.com/MichaIng/DietPi/pull/4411/files) 😉.

Joulinar replied:
Which one you suggest?

MichaIng replied:
`126` became free due to LibSSL1.0.0 removal.

Joulinar replied:
done

Joulinar committed:
v7.3

DietPi-Software | AdGuardHome - adjust product name
DietPi-Software | AdGuardHome - change software ID to 126
DietPi-Services | AdGuardHome - adjust product name

Joulinar committed:
v7.3

README.md | AdGuard Home - add new software title

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - change configuration file ID
DietPi-Software | AdGuard Home - adjust service configuration file

Joulinar committed:
Update AdGuardHome.yaml

AdGuardHome.yaml | remove non-default DNS filter lists

Joulinar committed:
Update dietpi-software

DietPi-Software | AdGuard Home - remove logfiles on uninstall from /var/log/

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8966,6 +9010,43 @@ _EOF_
+			fi
+
+			# Service
+			cat << _EOF_ > /etc/systemd/system/AdGuardHome.service
```
Let's not use capital letters for the service name. Probably it works, but it is very uncommon.

I would vote for this for all directories as well. We have a few other cases where capital letters are used for install directories, but it makes console navigation nasty (tab-completion by default is case-sensitive). Generally I like the consistency where service user = service name = install dir = data dir, all in lower-case letters 😃.

Joulinar replied:
I know, I just tried to follow the official service naming as much as possible. ;)

Joulinar replied:
ok changed to `adguardhome.service`

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -8966,6 +9010,43 @@ _EOF_
+WorkingDirectory=/opt/AdGuardHome
+
+StandardOutput=file:/var/log/AdGuardHome.out
+StandardError=file:/var/log/AdGuardHome.err
```
Let's keep STDOUT (default) as long log target, or do you think plain text log files have much benefit over `journalctl -u adguardhome`? The letter must be used anyway to check for systemd/service side failures and service state. This also avoids issues when using a non-root users (which I would prefer) where we'd need to have a subdirectory for log files pre-created.

Joulinar replied:
ok I removed the 2 lines. Output is going to STDOUT now

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - Fixes

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9018,7 +9018,7 @@ _EOF_
 			if [[ ! -f '/opt/AdGuardHome/AdGuardHome.yaml' ]]; then
 				dps_index=$software_id Download_Install 'AdGuardHome.yaml' /opt/AdGuardHome/AdGuardHome.yaml
-				G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 "" $GLOBAL_PW | tr -d ':\n' | sed 's/$2y/$2a/')" /opt/AdGuardHome/AdGuardHome.yaml
+				G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 "" $GLOBAL_PW | tr -d ":\n" | sed "s/$2y/$2a/")" /opt/AdGuardHome/AdGuardHome.yaml
```
Hehe that fix was wrong. Now the shell tries to expand `$2`, which was never set, hence two empty strings are passed there. The single quotes were perfectly fine, which tell the shell to interpret all contained characters literally. To address the shellcheck warning and also avoid that `sed` interprets `$` as "end of line" (special regular expressions meaning, although in this case `sed` is smart enough), you can escape it with a backslash to tell everyone that it is meant literal:
```
G_CONFIG_INJECT 'password:[[:blank:]]' "  password: $(htpasswd -bnBC 10 '' '$GLOBAL_PW' | tr -d ':\n' | sed 's/\$2y/\$2a/')" /opt/AdGuardHome/AdGuardHome.yaml
```
The single quotes around `$GLOBAL_PW` are important to avoid issues with special characters and spaces in the password. The variable is still expanded as the whole thing is double quoted. So the shell first expands `$GLOBAL_PW` and then passes it together with the single quotes to the `$(...)` command substitution which hence has a single-quoted literal password string. The more I write to more I see that it's quite complicated with nested arguments, command substitutions and commands which all have their syntax and special character interpretation rules 😄.

Joulinar replied:
this is not working as well 😉 
single quotes around `$GLOBAL_PW` will set `$GLOBAL_PW` as password. The word and not the variable 🤣

MichaIng replied:
An lol, true, as it's inside of the command substitution. Okay then use double quotes, around it.

Joulinar replied:
😟 still complaining

MichaIng replied:
I disabled the check for this line now. I though I solved it in other cases inside `sed` with a backslash already 🤔: https://github.com/koalaman/shellcheck/wiki/SC2016
> In the particular case of `sed`, ShellCheck uses additional heuristics to try to separate cases like `'s/$foo/bar/'` (failing to replace the variable `$foo`) with from the false positives like `'$d'` (delete last line).

Not so smart, as a backslash is THE sign for "take it literal" 😉. I'll open an issue there, when I find time to search through all the existing SC2016 false alarm reports that exist already 😄.

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - rename service and keep output on STDOUT (default)
DietPi-Services | AdGuard Home - rename service

Joulinar committed:
v7.3

AdGuardHome.yaml | change default port to 8083

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - code fixe

MichaIng committed:
v7.2

+ DietPi-Software | Satisfy shellsheck: Shouldn't be SC2016 satisfied automatically when $ signs are already escaped explicitly? :(

Joulinar reviewed (commented):
This is close to cheating 😉

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - remove environment configuration from service description

MichaIng committed:
v7.3 (#4440)

META | Update AdGuardHome branch

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - Avoid having AdGuard Home installed in parallel to Pi-hole
DietPi-Software | AdGuard Home - Offer to install Unbound along with AdGuard Home

Joulinar committed:
v7.3

code fix, remove break

Joulinar committed:
v7.3

DietPi-Software | Unbound - configure Unbound to work together with AdGuard Home

Joulinar committed:
v7.3

DietPi-Software | Unbound - Configuring AdGuard Home to use Unbound

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - Unbound: Switch port to 5335 if it was installed before
DietPi-Software | AdGuard Home - change configuration to use Unbound if available
DietPi-Software | Unbound - Assure AdGuard Home does not resolve via Unbound anymore if Unbound will be uninstalled
DietPi-Software | Pi-hole – on uninstallation, switch Unbound port to 53 if Unbound is still installed
DietPi-Software | AdGuard Home – on uninstallation, switch Unbound port to 53 if Unbound is still installed

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - ensure no concurrent installation of AdGuard Home and Pi-hole, even if installation is started from CLI

Joulinar committed:
v7.3

DietPi-Software | Unbound - Show a notification only, instead of a message pop-up if Unbound is being uninstalled

MichaIng committed:
v7.3

+ README | Fix AdGuard Home link

MichaIng committed:
v7.3

+ DietPi-Software | AdGuard Home: Do Pi-hole concurrency check as part of the dependency function, like we do with Tor Relay and Tor Hotspot. This is called on CLI installs as well while on menu-based installs the additional warnings are great so that users have chance to adjust their selections before starting the installs.
+ DietPi-Software | AdGuard Home: Assure that file modes are kept when copying files to disk and remove files from tmpfs afterwards
+ DietPi-Software | AdGuard Home: Only enable service during install and start it after all installs have been finished, to keep RAM and CPU usage at a minimum. If it was installed already, the service is still running, which is then required to avoid stopping the systems own DNS server.
+ DietPi-Software | Assure that uninstalls do not fail uncleanly when single steps fail, but keep the interactive error prompt. This way it is made clear that something failed, but instead of "Exit", "Ignore" can be selected to proceed with the uninstall.

MichaIng committed:
v7.3

+ DietPi-Software | Minor coding and remove trailing spaces

MichaIng committed:
v7.3

+ DietPi-Software | Pi-hole: Remove obsolete Unbound config removal, as this file gets removed on DietPi-Update v6.34 => v7.0 patches and hence cannot exist anymore

Joulinar committed:
v7.3

DietPi-Software | Lighttpd - stop service before uninstalling, otherwise an active process will be left
DietPi-Software | Lighttpd – remove enabled configuration directory, to avoid other software title trying to disable Lighttpd mods
DietPi-Software |Pi-hole – disable Lighttpd mods one by one as using wildcards does not seems to be working
DietPi-Software |Pi-hole – correct folder name for available Lighttpd mods to allow correct removal during uninstallation

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -13213,8 +13213,14 @@ _EOF_
 			G_AGP lighttpd
 
+			# Remove enabled Lighttpd configuration directory
+			[[ -d '/etc/lighttpd/conf-enabled' ]] && rm -R /etc/lighttpd/conf-enabled
```
Interesting. Probably we should remove the whole `/etc/lighttpd` then. E.g. on Pi-hole uninstall `/etc/lighttpd/conf-available` is checked, so that should be avoided then as well.

Joulinar replied:
Was thinking the same and this was my first try but changed it to the sub directory to keep user created configuration files if available. But we could remove the entire folder. Just need to agree on 😉

MichaIng replied:
At least removing the whole dir is how we do it in other cases as well. Uninstalling software basically means purging data and config, which is also stated before confirming the uninstall. A differentiation between "uninstall"/"remove" and "purge", like APT has it, would be another nice feature for the future, but until then, I prefer to remove things cleanly 🙂.

Joulinar replied:
Ok I will change to have the entire folder removed

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -14180,10 +14186,10 @@ _EOF_
 			fi
 			# - Lighttpd
-			if [[ -d '/etc/lighttpd/mods-available' ]]; then
+			if [[ -d '/etc/lighttpd/conf-available' ]]; then
```
Good find 👍.

`lighty-disable-mod` fails with exit code when the config is disabled already. So to be sure, and to allow printing error messages (removing `2> /dev/null`) and principally error handling that call, we can check for the enabled config explicitly:
```
[[ -f '/etc/lighttpd/conf-enabled/99-dietpi-pihole.conf' ]] && G_EXEC_NOEXIT=1 G_EXEC lighty-disable-mod dietpi-pihole
[[ -f '/etc/lighttpd/conf-enabled/99-dietpi-pihole-block_public_admin.conf' ]] && G_EXEC_NOEXIT=1 G_EXEC lighty-disable-mod dietpi-pihole-block_public_admin
G_EXEC_NOEXIT=1 G_EXEC rm -f /etc/lighttpd/conf-{available,enabled}/99-dietpi-pihole{,-block_public_admin}.conf
```
Actually... `lighty-disable-mod` does nothing else than removing the symlink, so the last command covers the first two. Usually I like to use the CLI, to be sure that all is done correctly, but in this case it is very trivial, and if something did change in Lighttpd/with these commands, then we'd need to change a lot of scripts anyway. What do you think? Keeping it simple and remove configs without relying on those CLI commands?
```
G_EXEC_NOEXIT=1 G_EXEC rm -f /etc/lighttpd/conf-{available,enabled}/99-dietpi-pihole{,-block_public_admin}.conf
```

Joulinar replied:
Yes we could check one by one but I'm more in favour of a simple one-liner. At the end it is an uninstall and are going to clean it up anyway.

MichaIng replied:
I mean it's two configs only. But yes the more I think about it the less I see a reason to go the complicated way. The last command will remove both configs, enabled symlink and available file, regardless if Lighttpd binaries are actually still installed or only the `/etc/lighttpd` directory structure left, so that is all we need. With `-f` it ignores missing files without error out, so no need to check for each file/link individually.

Joulinar committed:
v7.3

DietPi-Software | Lighttpd - enhanced Lighttpd uninstall process

Joulinar committed:
v7.3

DietPi-Software | Lighttpd - use correct path to determine if lighttpd service is available on uninstallation

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - Allow AdGuardHome to run without superuser privileges

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - change permission on home and working directory

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - Allow AdGuardHome to run without superuser privileges via service value `AmbientCapabilities`

Joulinar committed:
v7.3

AdGuardHome.yaml | reduce query log interval down to 7 days

Joulinar committed:
v7.3

DietPi-Software | AdGuard Home - consequently have identical lower-case name for user, service and directories
DietPi-Software | AdGuard Home – change working directory to dietpi_userdata
DietPi-Software | AdGuard Home – move config files and data directory to dietpi_userdata
DietPi-Software | AdGuard Home – reduce permission of install directory /opt/adguardhome to 755
DietPi-Software | AdGuard Home – allow DHCP server scanning by setting CAP_NET_RAW within service file
DietPi-Software | AdGuard Home – add reload option within service file

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -4394,6 +4423,56 @@ amvdec_vp9' > /etc/modules-load.d/dietpi-c4-kodi.conf
+			[[ -d '/opt/adguardhome' ]] && rm -R /opt/adguardhome
+
+			G_EXEC cp -a AdGuardHome /opt/adguardhome
+			G_EXEC_NOFAIL=1 G_EXEC rm -R AdGuardHome
```
`G_EXEC mv AdGuardHome /opt/adguardhome` is a bit simpler 🙂.

Joulinar committed:
v7.3

smaller change

MichaIng commented on dietpi/dietpi-software:
```diff
@@ -9055,33 +9056,34 @@ _EOF_
 			# Permissions
-			chown -R adguardhome:adguardhome /mnt/dietpi_userdata/adguardhome /opt/AdGuardHome/
+			chown -R adguardhome:adguardhome /mnt/dietpi_userdata/adguardhome /opt/adguardhome
+			chmod 755 /opt/adguardhome{,/AdGuardHome}
```
We could use `G_EXEC` here and in a few more cases.

Joulinar replied:
😃

Joulinar replied:
probably to much now 🙈

MichaIng replied:
Actually there is no reason to not use it everywhere, where no redirect is involved (which is difficult to pass to a function). That way it's also more transparent what we're actually doing during the installs. All fine for me. On uninstalls I used to use `G_EXEC_NOEXIT=1` currently to allow ignoring errors, but in the cases here, there would be a critical issues present if any of those fails, so IMO fine to stay strict.

Joulinar committed:
v7.3

smaller cosmetics

MichaIng committed:
v7.3

+ CHANGELOG | AdGuard Home: This DNS sinkhole ad blocker for your LAN, similar to Pi-hole, has been added with software ID 126.
